crates/datafusion_ext/src/metrics.rs

```diff
@@ -248,6 +256,75 @@ impl<T: DataSourceMetricsOptsType> ExecutionPlan for DataSourceMetricsExecAdapte
         self: Arc<Self>,
         children: Vec<Arc<dyn ExecutionPlan>>,
     ) -> Result<Arc<dyn ExecutionPlan>> {
+        if children.is_empty() {
+            return Err(DataFusionError::Plan(
+                "DataSourceMetricsExecAdapter wrong number of children".to_string(),
+            ));
+        }
+        Ok(Arc::new(Self::new(children[0].clone())))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        let stream = self.child.execute(partition, context)?;
+        Ok(Box::pin(BoxedStreamAdapater::new(
+            stream,
+            partition,
+            &self.metrics,
+            WriteOnlyDataSourceMetricsOptsType::OPTS,
+        )))
+    }
+
+    fn statistics(&self) -> Result<Statistics> {
+        self.child.statistics()
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+
+    fn required_input_ordering(&self) -> Vec<Option<Vec<PhysicalSortRequirement>>> {
+        // the input ordering must match the output ordering of the source
+        // without this, the sort may get removed by the optimizer
+        vec![self
+            .output_ordering()
+            .map(PhysicalSortRequirement::from_sort_exprs)]
+    }
+}
```

Comment:
the `WriteOnlyDataSourceMetricsExecAdapter` needs an explicit check on the input ordering, otherwise the `EnforceSorting` rule may remove it. The `ReadOnly...` doesn't care, so this needed to be split out. 

This states that if the child has an explicit `output_ordering`, then this node needs a matching `input_ordering`.